### PR TITLE
venue type

### PR DESCRIPTION
**Pull request by radius314**
https://github.com/bmlt-enabled/bmlt-root-server/pull/370/files?diff=unified&w=1

### CODE REVIEW

**Comment by alanborning** [comment]
I think this may need some additional checks for bad data.  If users have created meetings using the new editor, things will be consistent and all will be well. But what if they checked an inconsistent set of formats using the old interface?  There are alas a lot of cases of that out there. Here are some problem cases I noticed:
- If TC is checked but not VM or HY in the old editor, the meeting shows as in-person in the new editor. TC is still in the formats list and will be exported, but you can no longer see it to remove it. This is still true if you make a change to some other format.
- If HY and VM are both checked, the meeting shows as HY in the new editor (but VM still gets exported). This is still true if you make a change to some other format. However, if you select a different venue type and then select hybrid again, the export is correct.
- If HY, TC, and VM are all checked, the meeting shows as hybrid.  Same as before regarding changes ....

For these cases we may want to flag the error and not let the user save changes to the meeting until it's fixed.

Finally, if VM is checked but not TC, but the venue still has a location etc, the meeting just shows as a pure virtual meeting. (Should there be a check for this one?  Seem less serious than the others.)

**Review by jbraswell — CHANGES_REQUESTED** [review]
Overall, I think this approach looks really good. I think I agree with Alan's comment, and made a suggestion for how to move forward. If we decide to move forward with the suggestion I made, I think the only thing missing is some input validation requiring the user to select at least one radio button before saving.

File: main_server/local_server/server_admin/server_admin_javascript.js
Comment (jbraswell) [reply main_server/local_server/server_admin/server_admin_javascript.js:4898]: If it turns out that someone out there has done something silly and deleted one of these formats, I don't mind adding a database migration to add it back. After it's added back, because we're disabling deletion of these going forward, we should be set.

**Comment by radius314** [comment]
> Overall, I think this approach looks really good. I think I agree with Alan's comment, and made a suggestion for how to move forward. If we decide to move forward with the suggestion I made, I think the only thing missing is some input validation requiring the user to select at least one radio button before saving.

A radio button is always selected currently.  Because if none of the 3 combinations are valid, `In Person` is automatically selected.  There is no way to unselect here.

File: main_server/local_server/server_admin/server_admin_javascript.js
Comment (radius314) [reply main_server/local_server/server_admin/server_admin_javascript.js:4898]: I think there was another ticket addressing this right?  https://github.com/bmlt-enabled/bmlt-root-server/issues/362

**Comment by jbraswell** [comment]
> > Overall, I think this approach looks really good. I think I agree with Alan's comment, and made a suggestion for how to move forward. If we decide to move forward with the suggestion I made, I think the only thing missing is some input validation requiring the user to select at least one radio button before saving.
> 
> 
> 
> A radio button is always selected currently.  Because if none of the 3 combinations are valid, `In-Person` is automatically selected.  There is no way to unselect here.

I know. I made a suggestion in another comment suggesting an alternate implementation. I even provided code.

File: main_server/local_server/server_admin/server_admin_javascript.js
Comment (jbraswell) [reply main_server/local_server/server_admin/server_admin_javascript.js:4898]: Yes, but that doesn't prevent someone from deleting them before they upgrade. I'm saying it's worth checking to see what folks have done, and I might need to write a migration to be doubly sure they exist depending on what we find.

Comment (radius314) [reply main_server/local_server/server_admin/server_admin_javascript.js:4898]: Ok I'm not sure how to handle this.

File: main_server/local_server/server_admin/server_admin_javascript.js
Comment (radius314) [reply main_server/local_server/server_admin/server_admin_javascript.js:2998]: What was the final decision on what you want implemented?

**Comment by radius314** [comment]
> > > Overall, I think this approach looks really good. I think I agree with Alan's comment, and made a suggestion for how to move forward. If we decide to move forward with the suggestion I made, I think the only thing missing is some input validation requiring the user to select at least one radio button before saving.
> > 
> > 
> > A radio button is always selected currently.  Because if none of the 3 combinations are valid, `In-Person` is automatically selected.  There is no way to unselect here.
> 
> Right. I made a suggestion in another comment suggesting an alternate implementation, and I provided code. The only thing missing from my example code is the extra input validation to require a selection.

I'm not following.  Unless I'm missing something, there would never be a scenario where a radio button would be unselected, because an option is always pre-selected and calculated.

File: main_server/local_server/server_admin/server_admin_javascript.js
Comment (jbraswell) [reply main_server/local_server/server_admin/server_admin_javascript.js:2998]: > > > > Overall, I think this approach looks really good. I think I agree with Alan's comment, and made a suggestion for how to move forward. If we decide to move forward with the suggestion I made, I think the only thing missing is some input validation requiring the user to select at least one radio button before saving.
> > > 
> > > 
> > > A radio button is always selected currently.  Because if none of the 3 combinations are valid, `In-Person` is automatically selected.  There is no way to unselect here.
> > 
> > 
> > Right. I made a suggestion in another comment suggesting an alternate implementation, and I provided code. The only thing missing from my example code is the extra input validation to require a selection.
> 
> I'm not following. Unless I'm missing something, there would never be a scenario where a radio button would be unselected, because an option is always pre-selected and calculated.

https://github.com/bmlt-enabled/bmlt-root-server/pull/370#discussion_r549157743

The code here will not always pre-select a radio button.

Comment (jbraswell) [reply main_server/local_server/server_admin/server_admin_javascript.js:2998]: I think so? I don't know if the code works or not. The outcome should be that when a supported combination of formats is not selected, no radio button is selected. Then, the user should not be able to save without making a selection, forcing them to fix their data before they can edit their meeting.

**Comment by alanborning** [comment]
The cases I was concerned about all arise if there were an inconsistent set of formats already selected using the old editor or the app. Yes, a radio button is always selected even in that case, but the set of formats might still be inconsistent. I had a list of cases I tried in the earlier comment. As one example, suppose both VM and HY are checked. My suggestion here is that the user be notified, and asked to make things consistent before proceeding. You can't save unless the formats are consistent. The system could make a suggestion about what its guess is -- in this case, it should guess just HY -- but the user would need to confirm.

An alternative would be to silently guess what was intended and save a consistent state (e.g., silently uncheck VM and leave HY). That seems more hazardous but would work.

Just showing the HY radio button pressed, but leaving both VM and HY checked, seems wrong.

**Comment by radius314** [comment]
I definitely want to improve how the validation works and not use alerts.  But this is the simplest implementation placeholder for now.

image

**Comment by jbraswell** [comment]
Sweet. Yeah, I'm with you on not using alerts, and agree we can do that later. This looks good.

I'll do a proper code review later tonight or tomorrow.

**Comment by alanborning** [comment]
I checked the behavior of the code on all the possible combinations of TC, HY, and VM - all work correctly!!

**Comment by alanborning** [comment]
Also checked the behavior for making new meetings of each possible kind, and NAWS export functionality for the different kinds of meetings.

**Review by jbraswell — APPROVED** [review]

**Comment by alanborning** [comment]
Minor code issue: the string `"You must specify a venue type.`" in server_admin_javascript.js should be localized.

**Comment by radius314** [comment]
> Minor code issue: the string `"You must specify a venue type.`" in server_admin_javascript.js should be localized.

this will get refactored as part of cleaning up the validation, I don't want to hold up merging it for that.

**Pull request merged**
